### ISSUE TITLE
[LDB-138] filtering returns incorrect count using 1M row table

### ISSUE THREAD

**Issue by dynamobi-build** [LucidDB/luciddb#183]
<b><i>[reporter="schoi", created="Tue, 12 Sep 2006 12:36:15 -0500 (GMT-05:00)"]</i></b>
To repro:<br/>
--------------<br/>
/open/luciddb/test/sql/sqbm$  ant test -Dbh.test-names=sqbm.sqbm2b<br/>
<br/>
It takes approx 4.5 minutes to run (load and multiple select count(_) from a 1 million row table)<br/>
<br/>
Fails at this query:<br/>
select count(_) from bench1m where k2 = 2 and not (K1K = 3);<br/>
<br/>
-- returns:<br/>
+---------+<br/>
| EXPR$0  |<br/>
+---------+<br/>
| 499897  |<br/>
+---------+<br/>
<br/>
-- expecting<br/>
+---------+<br/>
| EXPR$0  |<br/>
+---------+<br/>
| 499901  |<br/>
+---------+<br/>
<br/>
Note that I just updated the ref file .. the correct return should be 499901. (checked against excel)  Before, when this test was "passing", it was returning 499900.

**Comment by dynamobi-build** [comment]
<b><i>[author="jvs", created="Tue, 12 Sep 2006 12:39:29 -0500 (GMT-05:00)"]</i></b>
Off-topic:  running sqbm without indexes should be a good regression test for residual filters.

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Tue, 12 Sep 2006 17:05:18 -0500 (GMT-05:00)"]</i></b>
I need to do some more analysis, but what I've found so far either points to a problem in index search  for a filter like "not(k1k = 3)" or in LcsIndexMergeRel.  Here's what I did:<br/>
<br/>
create table k2res(rid bigint, k2 int, k1k int);<br/>
create index i on k2res(k1k);<br/>
insert into k2res select lcs_rid(kseq), k2, k1k from bench1m where k2 = 2;<br/>
<br/>
select count(_) from k2res where k1k <> 3;<br/>
-- returns correct count -- 499901<br/>
<br/>
select count(_) from k2res where not(k1k = 3);<br/>
-- returns incorrect count -- 499899<br/>
<br/>
The first query does not use an index, whereas the second one does.<br/>
 <br/>
I inserted the rids from the two queries into separate tables and then did an outer join on the two tables filtering out the rows that were null extended.  The two missing rids are 360427 and 659379.<br/>
<br/>
On another note,I tried getting the missing rids by doing an EXCEPT, but it returned no rows.  So, it looks like there's also a problem in EXCEPT.

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Wed, 13 Sep 2006 11:09:31 -0500 (GMT-05:00)"]</i></b>
Upon further investigation, it now looks like this may be a mergeEntry problem!  Expanding on what I did in the previous comment, I was able to track down that it appears to be the index search that is first losing one of the missing rids (360427).  In my new table, k2res, that rid value corresponds to the row with lcs_rid = 180430.  Tracing the index values stored in the index that's on k2res(k1k), shows that rid value 180430 is completely missing from that btree.<br/>
<br/>
So far, I've traced LbmGeneratorExecStream and LbmSplicerExecStream.  Generator is generating a bitmap entry with rid 180430, but splicer is not.  I suspect there may be a problem in mergeEntry handling a corner case because the missing rid (180430) is at the very end of the first entry to be spliced, and the next entry that it's being spliced with is a singleton with rid value 184230.  From the splicer trace, it looks like the attempt to splice in 184230 runs out of space, and mergeEntry decides to produce the first entry.  But when it does, it loses the last rid in the entry.  I need to debug further to confirm this.<br/>
<br/>
I'm guessing this behavior is now showing up because of the changes I made for the resource governor.  With the new resource governor, generator is getting allocated more scratch pages.  With more scratch pages, generator is likely generating bigger entries, which changes the entries that end up getting spliced.

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Wed, 13 Sep 2006 12:49:09 -0500 (GMT-05:00)"]</i></b>
So, the problem does appear to be a space accounting problem in mergeEntry.  In mergeEntry, getMergeSpaceRequired() is called to get the space requirements for the new tuple to be merged.  However, the space requirements do not take into account space for zero bytes. <br/>
<br/>
Here are the numbers for the smaller repro I was running.<br/>
<br/>
- An entry with startRID 167032 gets merged into the current entry.  The size of the current entry is 469.<br/>
- getMergeSpaceRequired() indicates that the new entry needs 52 bytes.  469+52= 521, which is <= scratchBufferUsableSize, which is also 521.<br/>
- However, the new entry is further away from the current entry so zero bytes need to be added to the last segment in the current entry.  In this case, 2 bytes.  So, once the new entry is merged into the currentEntry, currentEntrySize = 469+2+52 = 523.<br/>
- The next entry that comes in to be merged also is far away from the new current entry and therefore, it too needs zero bytes.  It needs 2 as well.<br/>
- 523+2 = 525, which exceeds scratchBufferSize!  Consequently, later when the final segment descriptor with the 2 byte zero byte length is written, it overwrites the last byte segment.<br/>
  <br/>
  So, basically, as I see it, the problem is when you're reserving space for the zero length bytes, you're reserving enough space for only 1 descriptor's zero bytes, and not 2, which is needed in the case.<br/>
  <br/>
  To verify this is the problem, I changed getMergeSpaceRequired to add on LbmZeroLengthExtended.  With that, the reduced testcase passes.<br/>
  <br/>
  I don't know if this is the real fix, as you may want to optimize by only adding on the extra zero length space in getMergeSpaceRequired, if it's really needed.<br/>

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Wed, 13 Sep 2006 12:54:58 -0500 (GMT-05:00)"]</i></b>
Rushan -- I need to look at another performance issue, so since I've isolated the root cause, hopefully, it'll be straightforward for you to fix this one :-).<br/>
<br/>
I have not retried my temporary fix on the larger dataset so I don't know if there are other problems lurking in the corners.  (Hopefully not.)  Just a warning, you should build an optimized build to do the data load.  Sunny's 4.5 minute number to repro the problem is based on an optimized build.  When I tried loading using a debug build, it didn't finish after almost an hour so I aborted it and switched to an optimized build.  And then went back to a debug build to debug the problem, once I had loaded the 1 million row table and created the smaller test table k2res.

**Comment by dynamobi-build** [comment]
<b><i>[author="rchen", created="Mon, 18 Sep 2006 11:15:56 -0500 (GMT-05:00)"]</i></b>
This problem is caused by not reserving enough space when inserting singletons. The fix is to change the order of the growEntry() call which reserves space for incoming bitmap tuple, and the stRID() call which merges(by inserting the RID) an singleton bitmap tuple with the current entry. Soem adjustments also need to be made in openLastSegment and growEntry to accept the singleton case.<br/>
<br/>
Also fixed is the other bug discovered above regarding EXCEPT operator using hash join. This bug existed before and was not uncovered because test data sets are normally smaller than the hash table size. When number of distinct keys is large, several keys are stored in the same slot of the hash table. Some of these keys might not match during join(or set op). These keys need to be returned as result of anti-join(which implements EXCEPT). The bug is to check all keys stored in a slot for matching indicator not just the first key in the slot.<br/>
<br/>
A note on reproducing the problem:<br/>
<br/>
It takes a long time to build the indexes associated with bench1m. Creating k2res directly from the flat file source bench_source is fast even in debug compile.<br/>
<br/>
create table k2res(rid bigint, k2 int, k1k int); <br/>
insert into k2res select lcs_rid(c1), c2, c1k from bench_source where c2 = 2; <br/>
<br/>
create index i on k2res(k1k); <br/>
<br/>
To verify the except fix works. These are the test cases used:<br/>
<br/>
create table k2copy(rid bigint, k2 int, k1k int);<br/>
nsert into k2res select lcs_rid(c1), c2, c1k from bench_source where c2 = 2 and c1 <> 100;<br/>
<br/>
select rid from k2res except select rid from k2copy;<br/>
<br/>

**Comment by dynamobi-build** [comment]
<b><i>[author="rchen", created="Mon, 18 Sep 2006 11:53:14 -0500 (GMT-05:00)"]</i></b>
Fix in eigenbase change list 7692.
